### PR TITLE
Refactor user registration

### PR DESCRIPTION
**Pull request by allenwq**
Follow #1749 

* Add EnrolRequests to handle user requests
* Remove workflow state of the CourseUser
* Delete TODO items when course_user is destroyed ( this fixes a bug that user cannot be enrolled again after un-enrolment )

Todo:
* Guide the user to create the account after the user clicks the confirmation link

### CODE REVIEW

File: app/services/course/user_registration_service.rb [hunk 1]
```diff
@@ -161,25 +137,17 @@ def invalid_code(registration)
   def accept_invitation(registration, invitation)
     CourseUser.transaction do
       invitation.confirm!
-      register_course_user(registration, workflow_state: :approved)
+      registration.course_user =
+        CourseUser.find_or_create_by!(course: registration.course, user: registration.user)
```
Comment (kxmbrian) [inline app/services/course/user_registration_service.rb]: Can do ` find_or_create_course_user!(registration)`

Comment (allenwq) [inline app/services/course/user_registration_service.rb]: 👍

**Comment by kxmbrian** [comment]
Can remove `:workflow_state` from `Course::UsersControllerManagementConcern#course_user_params` also

**Comment by kxmbrian** [comment]
So much simpler now. Looks good to me.

**Comment by allenwq** [comment]
@kxmbrian Thanks ! I removed the last place of `workflow_state` and fixed the hound complains.

**Pull request merged**